server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -759,7 +759,7 @@ private void allocateUnassigned() {
             final PriorityComparator secondaryComparator = PriorityComparator.getAllocationComparator(allocation);
             final Comparator<ShardRouting> comparator = (o1, o2) -> {
                 if (o1.primary() ^ o2.primary()) {
-                    return o1.primary() ? -1 : o2.primary() ? 1 : 0;
```

Comment:
why is this ok?

Reply:
`o1.primary() ^ o2.primary() && o1.primary() == false` implies that `o2.primary == true`